### PR TITLE
[Use Case]: create array specification use case

### PR DESCRIPTION
**Pull request by tmakar**
## Basics

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

**Comment by tmakar** [comment]
When clicking on  to check https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/PR-4811/4/pipeline to check what failed in the pipeline every step seems to be green but it still tells that there is an error. How is this possible?

**Comment by kodebach** [comment]
> How is this possible?

See https://github.com/ElektraInitiative/libelektra/issues/4671. You need to click "Tests" at the top right, to see what tests actually failed.

File: doc/usecases/plugins/spec/create_array_specification.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# Use Case: Create array specification for an application
+
+## Summary
+
+- **Scope:** `spec`
+- **Level:** Developer Goal
+- **Actors:** Application, User
```
Comment (markus2330) [inline doc/usecases/plugins/spec/create_array_specification.md:7]:
```suggestion
- **Actors:** Application Developer
```

File: doc/usecases/plugins/spec/create_array_specification.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+- **Brief:**
+  - Creating a specification for a REST application which needs communication to a mail server and message broker.
```
Comment (markus2330) [inline doc/usecases/plugins/spec/create_array_specification.md:9]: What you give here is an example, not a brief description of the use case.

File: doc/usecases/plugins/spec/create_array_specification.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+## Scenarios
+
+- **Precondition:** kdb is installed.
+- **Main success scenario:**
+  - Specification is written user or system-wide.
```
Comment (markus2330) [inline doc/usecases/plugins/spec/create_array_specification.md:15]: Please add more details about "is written", as it is it doesn't necessarily contain arrays. Verb after "system-wide" missing?

File: doc/usecases/plugins/spec/create_array_specification.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+meta:/type = string
+meta:/description = "The url for the mail server."
+
+[messagebroker/port]
+meta:/type = unsigned_short
+meta:/description = "The port for the message broker."
+
+[messagebroker/url]
+meta:/type = string
+meta:/description = "The url for the message broker."
```
Comment (markus2330) [inline doc/usecases/plugins/spec/create_array_specification.md:45]: Keep example minimal.

```suggestion
```

File: doc/usecases/plugins/spec/create_array_specification.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+service/#0/port = 587
+service/#0/url = tcp://localhost
+
+service/#1/port = 61616
+service/#1/url = tcp://localhost
+```
+
+We can see that this kind of specification is much better when it commes to having many services and similar configuration for the keys.
+But we can also identify a shortcoming with this `array specification` approach.
+We can no longer no what this services are i.e. if its a `mail server` or `message broker`.
```
Comment (markus2330) [inline doc/usecases/plugins/spec/create_array_specification.md:79]: Sentence?

File: doc/usecases/plugins/spec/create_array_specification.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+We have two solutions to this problem:
+
+1. Introducing another key `name` which identifies the service.
+2. Use another wildcard character `_` (see `create_specification_with_underline_wildcard`)
```
Comment (markus2330) [inline doc/usecases/plugins/spec/create_array_specification.md:84]: This would be a decision, it shouldn't be in a use case.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thank you for writing the use case. Something went wrong here: the title of the use case, brief and main success scenario describe totally different things.

**Comment by markus2330** [comment]
Btw. it is probably better if you use an example where arrays are perfectly suited, i.e., something where no name is needed.

**Review by kodebach — CHANGES_REQUESTED** [review]
Same as in #4809. The main scenario doesn't really describe what the use case is about. You should incorporate more of the example.

I also think you should leave the part about `name` and `_` out of this use case and create a separate use case for it. Here you can just use the mail client example again. Just start the scenario with:

```markdown
- **Main success scenario:**
  - The developer wants to extend their application from ["Writing specification for an application"](/doc/usecases/plugins/spec/create_specification.md) to support multiple mail account.
```

Then describe what is needed to make that work, i.e. switch from `mailserver/hostname` to `mailserver/#/hostname` etc.

File: doc/usecases/plugins/spec/create_array_specification.md (hunk 1 above)
Comment (tmakar) [reply doc/usecases/plugins/spec/create_array_specification.md:7]: Thanks!

File: doc/usecases/plugins/spec/create_array_specification.md (hunk 6 above)
Comment (tmakar) [reply doc/usecases/plugins/spec/create_array_specification.md:84]: Thanks!

File: doc/usecases/plugins/spec/create_array_specification.md (hunk 5 above)
Comment (tmakar) [reply doc/usecases/plugins/spec/create_array_specification.md:79]: Thanks!

File: doc/usecases/plugins/spec/create_array_specification.md (hunk 2 above)
Comment (tmakar) [reply doc/usecases/plugins/spec/create_array_specification.md:9]: Thanks!

**Comment by tmakar** [comment]
@markus2330 @kodebach 

Would be good if you find some time to review this so I can merge it 👍

**Review by kodebach — APPROVED** [review]
Same as in #4809. "Main success scenario" could explain the specification more instead of just explaining the configuration.

**Comment by tmakar** [comment]
@markus2330 also ready!

**Pull request closed without merging**